### ISSUE TITLE
[4.x]: Saving a matrix fields always changes all UIDs

### ISSUE THREAD

**Issue by MoritzLost** [craftcms/cms#11248]
### Description

Whenever I save a Matrix field, it generates a new UID for _every_ custom field in _every_ block type and super table type. This results in a ton of changes in the project-config, making PR reviews much harder and causing lots of merge conflicts.

Here's what saving a matrix field looks like, even if no changes are made:

```
❯ git status
On branch remove-unused-features
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   config/project/matrixBlockTypes/accordion--42e3ce00-cb9e-48db-b1ac-f1352e8b70e5.yaml
        modified:   config/project/matrixBlockTypes/banner--001021f0-ade8-4ecc-aa6f-36a266a0da11.yaml
        modified:   config/project/matrixBlockTypes/contact_form--0e7692e9-76f2-4c36-b035-0658ab05c677.yaml
        modified:   config/project/matrixBlockTypes/cta--79a16b14-e1a2-46f8-9892-8f0b9cef31a8.yaml
        modified:   config/project/matrixBlockTypes/downloads--3b550985-9e20-4e86-848b-8c5073395ffa.yaml
        modified:   config/project/matrixBlockTypes/gallery--435483f9-3f60-462e-b138-f81f1a4bcd02.yaml
        modified:   config/project/matrixBlockTypes/image--fa6e38c8-67d5-48e3-ab8b-8bece72c3256.yaml
        modified:   config/project/matrixBlockTypes/infobox--001eb95f-9896-41d8-b370-065cb32a9d24.yaml
        modified:   config/project/matrixBlockTypes/person--d711866f-76f7-4143-ad9f-cefa742943de.yaml
        modified:   config/project/matrixBlockTypes/section_start--3072d916-3cb4-4081-ba08-d8ae1c931f49.yaml
        modified:   config/project/matrixBlockTypes/statement--470fb8b4-e016-46d5-b853-6bad27d5633b.yaml
        modified:   config/project/matrixBlockTypes/teaser--604beba6-c11d-4d98-a2f2-0221207b0309.yaml
        modified:   config/project/matrixBlockTypes/text--626b4347-4a7f-4e6d-b1ae-4c86629728fe.yaml
        modified:   config/project/matrixBlockTypes/text--8418677c-7f60-43fe-b903-281edbc9ada0.yaml
        modified:   config/project/matrixBlockTypes/video_embed--124b34b9-d65e-4880-830c-c39e28186ffb.yaml
        modified:   config/project/matrixBlockTypes/video_local--84616049-5c32-4bc4-81f8-8da0ca69ef55.yaml
        modified:   config/project/project.yaml
        modified:   config/project/superTableBlockTypes/29d277ac-26b7-4476-bd6d-6522b88879e3.yaml
        modified:   config/project/superTableBlockTypes/a6be3a09-0acc-4769-b676-22f5c238ef6c.yaml
```

All the changes look like this:

```diff
--- a/config/project/matrixBlockTypes/accordion--42e3ce00-cb9e-48db-b1ac-f1352e8b70e5.yaml
+++ b/config/project/matrixBlockTypes/accordion--42e3ce00-cb9e-48db-b1ac-f1352e8b70e5.yaml
@@ -11,7 +11,7 @@ fieldLayouts:
             required: false
             tip: null
             type: craft\fieldlayoutelements\CustomField
-            uid: 2ac7cdc7-5afa-475a-89fa-89846f0e8e7c
+            uid: 18eebeb4-5d56-44d0-8a99-77636131d61e
             warning: null
             width: 50
           -
@@ -21,7 +21,7 @@ fieldLayouts:
             required: true
             tip: null
             type: craft\fieldlayoutelements\CustomField
-            uid: 5d4c4887-d72a-45b1-96a5-1f63871957c9
+            uid: a4c39f34-5fd7-4c10-9b0a-0e66149fd877
             warning: null
             width: 25
           -
@@ -31,7 +31,7 @@ fieldLayouts:
             required: false
             tip: null
             type: craft\fieldlayoutelements\CustomField
-            uid: a4d9713c-f355-4586-af09-9fc4a18fd6da
+            uid: 66b926da-3be9-49fd-8630-307502b22bec
             warning: null
             width: 25
```

Just commiting those changes once doesn't help, because all UIDs are replaced every time the field is saved. This didn't use to happen in Craft 4. Is this intended behaviour or a bug?

### Steps to reproduce

1. Save a matrix field with a couple of block types and fields in the backend.
2. A lot of changes will appear in the project config. Use `git add .` to stage them.
3. Save the field again. `git status` will show that all the UIDs for all fields in all matrix block types have changed again.

### Expected behavior

UIDs shouldn't change whenever the section is saved. This used to be the case in Craft 3.

### Actual behavior

In Craft 4, every time the field is saved a new set of UIDs is generated for no apparent reason.

### Craft CMS version

4.0.2

### PHP version

8.1.3

### Operating system and version

macOS 12.2.1

### Database type and version

MySQL 8.0

### Installed plugins and versions

- Super Table 3.0.0-beta.3

**Comment by brandonkelly** [comment]
Not able to reproduce that on my end. Can you try running `php craft project-config/rebuild` and see if that solves it?

**Comment by mmikkel** [comment]
I'm also seeing this and I'm able to consistently reproduce on a fresh Craft 4.0.2 install with no plugins (PHP 8.1.4, MySQL 5.7.37).

To be clear, the UIDs that change every time the Matrix field is re-saved are the field layout element UIDs, i.e. this:

![CleanShot 2022-05-19 at 09 27 54](https://user-images.githubusercontent.com/298510/169236902-0de14da4-2f01-414e-8af1-d4fad90d4f07.png)

From what I can tell, Matrix block field layout elements doesn't have UIDs on Craft 3.

**Comment by MoritzLost** [comment]
Thanks for the clarification @mmikkel, good to know it's not just us. We're seeing this on multiple unrelated projects that we recently updated to Craft 4, so it's unlikely something specific to one codebase or installation.

> Not able to reproduce that on my end. Can you try running php craft project-config/rebuild and see if that solves it?

@brandonkelly Same as @mmikkel, the command doesn't change any configuration save the project timestamp (we already did the rebuild during the initial Craft 4 update). After executing, we're still seeing the UID changes after every save.
